### PR TITLE
Colony Network working with a locked CLNY Token

### LINKED ISSUE #297: Working with a locked CLNY Token
With the locked [CLNY Token deployed to mainnet](https://github.com/JoinColony/colonyToken/blob/master/README.md#contracts) we need to test the functionality of the Meta Colony and Reputation mining against that. There will likely need to be adjustments in the `TokenAuthority` contract as to who can do token transfers under lock and to test for other blockers. 
Exact functionality that will need to work under locked CLNY is still tbc by @jakzilla

### PR DESCRIPTION
**Pull request by elenadimitrova**
Closes #297 

Integrates the `TokenAuthority` contract changes from [feature/upgrade-token-authority-for-colony-network](https://github.com/JoinColony/colonyToken/compare/feature/upgrade-token-authority-for-colony-network) branch in `colonyToken` repo. Those allow the Network to function with a locked CLNY token which the Colony Contribute release will have to work with initially.

I initially thought about keeping this functionality and testing separate from the final colony network release but in the end Colony Contribute will go live against the currently locked CLNY token on mainnet and so until that is unlocked we can either merge these changes or keep rebasing this branch to test how the network behaviour is affected by the locked CLNY.

Since the `Token` contract repeats itself between the two repositories, I have renamed the one here used by Colonies to create their token to `ERC20ExtendedToken` which better specifies its behaviour and left the Colony Token contract as `Token` which is how that is deployed to mainnet

Somewhat unrelated cleanup change here is to consolidate the various test folders into a common naming convention that groups them nicely in edit tree, e.g. `gasCosts` -> `test-gas-costs` and `upgrade-test` -> `test-upgrade`

**Test Accounts distribution**
accounts[0..2] are most commonly used as task roles.
accounts[3..10] are used as reputation miner accounts where accounts[5] is the goodClient (funded with default stake in migrations) and accounts[6,7] are bad clients.
accounts[11] is set as the CLNY Token owner

Creating and wiring up the MetaColony and CLNY is now consolidated in `setupMetaColonyWithLockedCLNYToken` and `setupMetaColonyWithUNLockedCLNYToken` to be used respectively in Colony Contribute functionality unit tests and post unlocking CLNY functionality, e.g. Token Auctions. 

Features NOT available with a locked CLNY:
- `ColonyNetworkAuction` creates a new instance of a `DutchAuction`  which does all CLNY transfers and needs to be authorised for that, and so do the users trading CLNY for Tokens.

### CODE REVIEW

File: test/colony-network-mining.js [hunk 1]
```diff
@@ -2254,7 +2277,7 @@ contract("ColonyNetworkMining", accounts => {
       this.timeout(100000000);
       const nClients = Math.min(accounts.length, 7);
       // TODO: This test probably needs to be written more carefully to make sure all possible edge cases are dealt with
-      for (let i = 0; i < nClients; i += 1) {
+      for (let i = 3; i < nClients; i += 1) {
```
Comment (area) [inline test/colony-network-mining.js]: This test needs more updating to restore its original behaviour.

File: helpers/test-helper.js [hunk 2]
```diff
@@ -414,13 +414,16 @@ export async function getActiveRepCycle(colonyNetwork) {
...
   await forwardTime(MINING_CYCLE_DURATION, test);
   const repCycle = await getActiveRepCycle(colonyNetwork);
+
   if (miningClient !== undefined) {
     await miningClient.submitRootHash();
   } else {
-    await repCycle.submitRootHash("0x00", 0, 10);
+    const accounts = await web3GetAccounts();
+    minerAddress = minerAddress || accounts[5]; // eslint-disable-line no-param-reassign
+    await repCycle.submitRootHash("0x00", 0, 10, { from: minerAddress });
```
Comment (kronosapiens) [inline helpers/test-helper.js]: Can we just default to `accounts[5]` and omit the argument? That's essentially what we had before when we defaulted to `accounts[0]`.

**Comment by kronosapiens** [comment]
Can you also talk a little more about the decision to move the mining responsibilities to the higher-numbered accounts? I see the benefit but it means we can't use the default `msg.sender == accounts[0]` anymore.

File: test/colony-network-mining.js [hunk 3]
```diff
@@ -214,11 +211,16 @@ contract("ColonyNetworkMining", accounts => {
...
+
         if (stakedBalance.gt(new BN(0))) {
           await tokenLocking.withdraw(clny.address, stakedBalance, { from: user });
         }
         const userBalance = await clny.balanceOf(user);
-        return clny.transfer(ZERO_ADDRESS, userBalance, { from: user });
+        if (userBalance.gt(new BN(0))) {
+          // CLNY transfers are locked, so approve the token owner to throw away the test accounts' tokens
+          await clny.approve(accounts[11], userBalance, { from: user });
+          await clny.transferFrom(user, ZERO_ADDRESS, userBalance, { from: accounts[11] });
```
[the remark this answers is not in the document]
Comment (elenadimitrova) [reply test/colony-network-mining.js]: Right

File: helpers/test-data-generator.js [hunk 4]
```diff
@@ -256,11 +273,10 @@ export async function setupFinalizedTask({
   });
 
   await colony.finalizeTask(taskId);
-
   return taskId;
 }
 
-export async function giveUserCLNYTokens(colonyNetwork, address, _amount) {
+export async function giveUserCLNYTokens(colonyNetwork, user, _amount) {
```
[the remark this answers is not in the document]
Comment (elenadimitrova) [reply helpers/test-data-generator.js]: Yes, but I thought you were going to remove the task flow here in your refactoring right?

File: helpers/test-helper.js (hunk 2 above)
Comment (elenadimitrova) [reply helpers/test-helper.js]: `repCycle` is a truffle contract and so we can't intercept this and have a "default". That plus I'd like us to understand who's calling this function.

File: helpers/test-data-generator.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kronosapiens) [reply helpers/test-data-generator.js]: Yup, just making sure.

File: parity-genesis.template.json [hunk 5]
```diff
@@ -46,9 +46,17 @@
...
-        "0000000000000000000000000000000deadbeef1": { "balance": "1000000000000000000000" },
-        "0000000000000000000000000000000deadbeef2": { "balance": "1000000000000000000000" },
-        "0000000000000000000000000000000deadbeef3": { "balance": "1000000000000000000000" },
-        "0000000000000000000000000000000deadbeef4": { "balance": "1000000000000000000000" }
+        "000000000000000000000000000000deadbeef10": { "balance": "1000000000000000000000" },
```
Comment (area) [inline parity-genesis.template.json]: Why start at deadbeef10?

File: test/colony-network-mining.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (area) [reply test/colony-network-mining.js]: So we can just burn them here, right? Or is burn locked?

File: helpers/test-helper.js (hunk 2 above)
Comment (area) [reply helpers/test-helper.js]: For the record, I think you can set defaults on `truffle-contract` instances, but I prefer making it explicit here too.

Comment (kronosapiens) [reply helpers/test-helper.js]: `Explicit is better than implicit`, or so they keep telling me.

File: parity-genesis.template.json (hunk 5 above)
Comment (elenadimitrova) [reply parity-genesis.template.json]: Because to be a valid hex it needs to be a certain length. Either deadbeef10 or we start at deadbeef1 and at 10 we drop a zero from the prefix.

File: test/colony-network-mining.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (elenadimitrova) [reply test/colony-network-mining.js]: I've changed it so it uses `burn`. That is open to everyone to burn their tokens
